File: packages/core/src/types/mcp.ts
import type { AgentType } from './agent';

/**
 * MCP Server configuration (Agent-specific)
 * MCP 配置直接属于 Agent，无全局注册表
 */
export interface AgentMCPServer {
  id: string;
  name: string;
  type: 'stdio' | 'sse' | 'http';
  enabled: boolean;
  connectionStatus?: 'connected' | 'disconnected' | 'error';
  config: MCPServerConfig;
  createdAt: string;
  updatedAt: string;
}

/** MCP Server config union type */
export type MCPServerConfig =
  | MCPStdioConfig
  | MCPSSEConfig
  | MCPHTTPConfig;

export interface MCPStdioConfig {
  type: 'stdio';
  command: string;
  args?: string[];
  env?: Record<string, string>;
  cwd?: string;
}

export interface MCPSSEConfig {
  type: 'sse';
  url: string;
  headers?: Record<string, string>;
}

export interface MCPHTTPConfig {
  type: 'http';
  url: string;
  headers?: Record<string, string>;
  method?: 'GET' | 'POST';
}

/**
 * Agent and its MCP servers (for frontend display)
 * 用于前端展示的 Agent 及其 MCP 配置列表
 */
export interface AgentMCPList {
  agent: AgentType;
  agentDisplayName: string;
  configPath: string;
  format: 'json' | 'toml' | 'yaml';
  servers: AgentMCPServer[];
}

/** MCP format mapping */
export interface MCPFormatMapping {
  agent: AgentType;
  format: 'json' | 'toml' | 'yaml';
  configPath: string;
  configKey: string;
}

/** Input for adding/updating MCP server */
export interface MCPServerInput {
  name: string;
  type: 'stdio' | 'sse' | 'http';
  config: MCPServerConfig;
  enabled?: boolean;
}

// Legacy types (kept for backward compatibility during migration)
// These will be removed after full migration

/** @deprecated Use AgentMCPServer instead */
export interface MCPServer {
  id: string;
  name: string;
  type: 'stdio' | 'sse' | 'http';
  enabled: boolean;
  connectionStatus?: 'connected' | 'disconnected' | 'error';
  config: MCPServerConfig;
  agents: MCPAgentBinding[];
  createdAt: string;
  updatedAt: string;
}

/** @deprecated No longer needed - MCP is directly tied to Agent */
export interface MCPAgentBinding {
  agent: AgentType;
  enabled: boolean;
  configPath: string;
  format: 'json' | 'toml' | 'yaml';
  syncStatus: 'pending' | 'synced' | 'error';
  lastSyncedAt?: string;
  lastError?: string;
}

/** @deprecated No longer needed - no global registry */
export interface MCPRegistryBinding {
  agent: AgentType;
  enabled: boolean;
  syncStatus: 'pending' | 'synced' | 'error';
  lastSyncedAt?: string;
  lastError?: string;
}

/** @deprecated Use direct agent parameter instead */
export interface MCPAgentBindingInput {
  agent: AgentType;
  enabled: boolean;
}

/** @deprecated No global registry anymore */
export interface MCPRegistryServer {
  id: string;
  name: string;
  type: 'stdio' | 'sse' | 'http';
  config: MCPServerConfig;
  agents: MCPRegistryBinding[];
  createdAt: string;
  updatedAt: string;
}

/** @deprecated No global registry anymore */
export interface MCPRegistryFile {
  version: 1;
  servers: Record<string, MCPRegistryServer>;
}

File: apps/desktop/src-tauri/src/commands/mcp.rs
use crate::error::CommandError;
use crate::services::mcp::MCPEngine;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MCPServer {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    pub server_type: String,
    pub enabled: bool,
    #[serde(rename = "connectionStatus")]
    pub connection_status: Option<String>,
    pub config: serde_json::Value,
    pub agents: Vec<MCPAgentBinding>,
    #[serde(rename = "createdAt")]
    pub created_at: String,
    #[serde(rename = "updatedAt")]
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MCPAgentBinding {
    pub agent: String,
    pub enabled: bool,
    #[serde(rename = "configPath")]
    pub config_path: String,
    pub format: String,
    #[serde(rename = "syncStatus")]
    pub sync_status: String,
    #[serde(rename = "lastSyncedAt")]
    pub last_synced_at: Option<String>,
    #[serde(rename = "lastError")]
    pub last_error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MCPAgentBindingInput {
    pub agent: String,
    pub enabled: bool,
}

#[tauri::command]
pub async fn list_mcp_servers(agent: Option<String>) -> Result<Vec<MCPServer>, CommandError> {
    MCPEngine::list_mcp_servers(agent.as_deref())
}

#[tauri::command]
pub async fn add_mcp_server(
    name: String,
    server_type: String,
    config: serde_json::Value,
    agents: Vec<MCPAgentBindingInput>,
) -> Result<MCPServer, CommandError> {
    MCPEngine::add_mcp_server(&name, &server_type, config, &agents)
}

#[tauri::command]
pub async fn update_mcp_server(
    id: String,
    name: Option<String>,
    config: Option<serde_json::Value>,
    agents: Option<Vec<MCPAgentBindingInput>>,
) -> Result<MCPServer, CommandError> {
    MCPEngine::update_mcp_server(&id, name.as_deref(), config, agents.as_deref())
}

#[tauri::command]
pub async fn remove_mcp_server(id: String) -> Result<(), CommandError> {
    MCPEngine::remove_mcp_server(&id)
}

#[tauri::command]
pub async fn toggle_mcp_agent(
    server_id: String,
    agent: String,
    enabled: bool,
) -> Result<(), CommandError> {
    MCPEngine::toggle_mcp_agent(&server_id, &agent, enabled)
}

#[tauri::command]
pub async fn import_mcp_from_agent(agent: String) -> Result<Vec<MCPServer>, CommandError> {
    MCPEngine::import_mcp_from_agent(&agent)
}
